contracts/contracts/rem.system/src/producer_pay.cpp

```diff
@@ -182,30 +182,70 @@ namespace eosiosystem {
    }
 
    using namespace eosio;
-   void system_contract::claimrewards( const name& owner ) {
-      require_auth( owner );
-
-      const auto& prod = _producers.get( owner.value );
-      check( prod.active(), "producer does not have an active key" );
 
-      check( _gstate.total_activated_stake >= min_activated_stake,
-                    "cannot claim rewards until the chain is activated (at least 15% of all tokens participate in voting)" );
+   void system_contract::claim_perstake( const name& guardian )
+   {
+      const auto& voter = _voters.get( guardian.value );
 
       const auto ct = current_time_point();
+      check( ct - voter.last_claim_time > microseconds(useconds_per_day), "already claimed rewards within past day" );
+
+      if (_gstate.perstake_bucket > 1'0000) { //do not distribute small amounts
+         int64_t total_per_stake_pay = 0;
+
+         auto sorted_voters = _voters.get_index<"bystake"_n>();
+         _gstate.total_guardians_stake = 0;
+         for (auto it = sorted_voters.rbegin(); it != sorted_voters.rend() && it->staked >= _gremstate.producer_stake_threshold; it++) {
+            if ( vote_is_reasserted( it->last_reassertion_time ) ) {
+               _gstate.total_guardians_stake += it->staked;
+            }
+         }
+
+         for (auto it = sorted_voters.rbegin(); it != sorted_voters.rend() && it->staked >= _gremstate.producer_stake_threshold; it++) {
+            if ( vote_is_reasserted( it->last_reassertion_time ) ) {
+               const int64_t per_stake_pay = _gstate.perstake_bucket * (double(it->staked) / double(_gstate.total_guardians_stake));
+               
+               _voters.modify( *it, same_payer, [&](auto &v) {
+                  v.pending_perstake_reward += per_stake_pay;
+               });
+               
+               total_per_stake_pay += per_stake_pay;
+            }
+         }
+
+         _gstate.perstake_bucket -= total_per_stake_pay;
+         check(_gstate.perstake_bucket >= 0, "perstake_bucket cannot be negative");
+      }
```

Comment:
I think this belongs to the delegatebw action. So the proportions are updated whenever any account stakes or unstakes tokens rather than everytime someone claims their reward.

Reply:
Also, in **to_rewards** we should additionally check **vote_is_reasserted** and if some Guardian failed to reassert, call **update_guardians_shares** again.